### DISCUSSION TITLE
TypeScript over Rust in frontend?

### DISCUSSION THREAD

**Discussion by LesnyRumcajs** [ChainSafe/forest-explorer#39] [General]
@hanabi1224 to clarify so that everyone is on the same page. Why we should introduce TypeScript here? In what areas? What are the benefits over pure Rust approach? What is the state of the Rust ecosystem here? What are the considerations for the team skillset?

cc: @lemmih

**Comment by lemmih** [comment]
I'm hugely in favor of keeping everything in Rust. Rust is sufficient for a low-complexity web app, and it is our area of expertise. Plus, we can reuse Forest code.

**Comment by hanabi1224** [comment]
I love the idea of hosting the server-side stuff on Cloudfare worker. My only concern is about building UI(html/js/css) stuff in Rust (with Leptos). 

I tried a basic local UI development scenario to hot reload a small UI change (rewording StateNetworkName to StateNetworkName2) by running `npx wrangler dev --live-reload`, it took ~1min for the page to reload on my laptop while it should take <1s with a typical javascript UI framework.  (I know wrangler uses server side rendering and it is possible to reduce the hot reload time by testing UI changes with client side rendering but the dev process is not obvious to me. Also given that the HTML change needs to be re-compiled into client WASM, I don't think <1s is even possbile)

I have not yet explored enough on Leptos so here I quote the AI generated answer to `Lepos vs react pros and cons` on Google.

https://www.google.com/search?q=leptos+vs+react+pros+and+cons

Leptos and React are both frameworks for building web applications, but they differ in their approaches and features: 
Leptos
A Rust-based framework that uses fine-grained reactivity to update the DOM when state changes. Leptos can build both server-side rendered (SSR) and client-side rendered (CSR) apps. Some pros of Leptos include: 
- Faster build times: Leptos' CSR apps have faster build times and a quicker iterative development cycle. 
- Serverless environments: Leptos can be deployed from serverless environments. 
- JavaScript Web APIs: Leptos supports standard Web APIs and the web_sys crate. 
React
A JavaScript framework that can be used for a variety of applications. Some pros of React include: 
- Easy to find developers: It's easy to find developers who are familiar with React. 
- Reusable components: Developers can reuse components. 
- Optimized for high performance: React is optimized for high performance. 
- Flexibility: React can be used to display a simple page or a view, or combined with other tools to build a complex application.

**Comment by lemmih** [comment]
Compilation time is an issue. Right now, we always build a complete bundle (server component and frontend component) in release mode and optimize the bundles with `wasm-opt.` Switching away from release mode and skipping the server-side component reduces compilation time from 18s to 2s on my Linux and 1.8s on my Mac.

**Comment by LesnyRumcajs** [comment]
We decided to move forward with all-in-Rust. If a need arises, we'll revisit this topic.
